pkg/cloudprovider/providers/azure/azure_storage.go

```diff
@@ -19,23 +19,48 @@ package azure
 import (
 	"fmt"
 
+	"github.com/Azure/azure-sdk-for-go/arm/storage"
+	"github.com/Azure/go-autorest/autorest/to"
 	"github.com/golang/glog"
 )
 
+const (
+	defaultStorageAccountType  = string(storage.StandardLRS)
+	fileShareAccountNamePrefix = "f"
+)
+
 // CreateFileShare creates a file share, using a matching storage account
 func (az *Cloud) CreateFileShare(name, storageAccount, storageType, location string, requestGB int) (string, string, error) {
 	var err error
 	accounts := []accountWithLocation{}
-	if len(storageAccount) > 0 {
-		accounts = append(accounts, accountWithLocation{Name: storageAccount})
-	} else {
-		// find a storage account
-		accounts, err = az.getStorageAccounts()
-		if err != nil {
-			// TODO: create a storage account and container
-			return "", "", err
+	if len(storageAccount) == 0 {
+		if location == "" {
+			location = az.controllerCommon.location
+		}
+		if storageType == "" {
+			storageType = defaultStorageAccountType
+		}
+		// find a storage account that matches storageType
+		accounts, err = az.getStorageAccounts(storageType)
+		if err != nil || len(accounts) == 0 {
+			storageAccount = generateStorageAccountName(fileShareAccountNamePrefix)
+			glog.V(2).Infof("azureFile - no storage account found, begin to create a new storage account %s in resource group %s, location: %s, storageType: %s",
+				storageAccount, az.controllerCommon.resourceGroup, location, storageType)
+			cp := storage.AccountCreateParameters{
+				Sku:      &storage.Sku{Name: storage.SkuName(storageType)},
+				Tags:     &map[string]*string{"created-by": to.StringPtr("azure-file")},
+				Location: &location}
+			cancel := make(chan struct{})
+
+			_, errchan := az.controllerCommon.cloud.StorageAccountClient.Create(az.controllerCommon.resourceGroup, storageAccount, cp, cancel)
+			err = <-errchan
+			if err != nil {
+				return "", "", fmt.Errorf(fmt.Sprintf("Failed to create storage account %s, error: %s", storageAccount, err))
+			}
 		}
 	}
+	accounts = append(accounts, accountWithLocation{Name: storageAccount, StorageType: storageType, Location: location})
```

Comment:
If storageAccount in input parameter is empty, and later az.getStorageAccount loads matching accounts, then here we're adding an account with empty name to the array, which does not make sense.